aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -226,7 +228,9 @@ Tensor add_sparse(const Tensor& self, const Tensor& other, Scalar alpha) {
   // TODO: Why?! Can't we just flip the order here...
   TORCH_CHECK(!(self.is_sparse() && !other.is_sparse()),
               "add(sparse, dense) is not supported. Use add(dense, sparse) instead.");
-  Tensor result = at::empty({0}, self.options());
+  auto commonDtype = at::result_type(self, other);
+  alpha_check(commonDtype, alpha);
+  Tensor result = at::empty({0}, self.options().dtype(commonDtype));
```

Comment:
I'm a little confused here -- it looks like add_out does type promotion, so why do we need to do type promotion here as well?

Reply:
With add_out the result tensor is specified (e.g. for in-place ops). The existing code here assumed the result would have the same dtype as `self`. If the dtype of the result tensor is specified, then result_type(self, other) has to be castable (`at::can_cast`) back to the dtype of the result tensor.